# 📊 Customer Churn Prediction: Exploratory Data Analysis

---

## Project Overview

**Business Problem:**  
Customer churn (when customers stop doing business with a company) is a critical concern for subscription-based businesses. Acquiring new customers is 5-25x more expensive than retaining existing ones. This project aims to:

1. **Understand customer behavior patterns** through data exploration
2. **Identify key factors** that contribute to customer churn
3. **Segment customers** into meaningful groups for targeted retention strategies
4. **Build predictive models** to identify at-risk customers before they churn
5. **Provide actionable insights** to reduce churn and increase revenue

---

## Dataset: Telco Customer Churn

**Source:** IBM Sample Data Sets  
**Context:** Telecommunications company customer data  
**Target Variable:** Churn (Yes/No) - Whether customer left within last month

**Feature Categories:**
- **Demographics:** Gender, SeniorCitizen, Partner, Dependents
- **Services:** Phone, Internet, Online Security, Tech Support, etc.
- **Account Information:** Contract type, Payment method, Tenure, Charges

---

## Objectives for This Notebook

1. ✅ Set up the environment and load required libraries
2. ✅ Download and load the Telco Customer Churn dataset
3. ✅ Perform initial data inspection and validation
4. ✅ Understand the structure and content of each feature
5. ✅ Identify data quality issues (missing values, data types, etc.)
6. ✅ Document initial observations for deeper analysis

---

**Expected Outcome:** A clean, loaded dataset ready for comprehensive exploratory data analysis in Phase 2.

---

## 1. Environment Setup

Import all necessary libraries and configure display settings for optimal data exploration.

In [1]:
# Data Manipulation and Analysis
import pandas as pd
import numpy as np

# Data Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Utilities
import warnings
import os
from pathlib import Path

# Configuration
warnings.filterwarnings('ignore')  # Suppress warnings for cleaner output
np.random.seed(42)  # Set random seed for reproducibility

# Pandas display options for better readability
pd.set_option('display.max_columns', None)  # Show all columns
pd.set_option('display.max_rows', 100)      # Show up to 100 rows
pd.set_option('display.width', None)        # Auto-detect display width
pd.set_option('display.precision', 2)       # 2 decimal places for floats

# Matplotlib/Seaborn styling for professional visualizations
plt.style.use('seaborn-v0_8-darkgrid')      # Clean, professional style
sns.set_palette('husl')                      # Colorblind-friendly palette
plt.rcParams['figure.figsize'] = (12, 6)    # Default figure size
plt.rcParams['font.size'] = 10              # Default font size
plt.rcParams['axes.titlesize'] = 14         # Title font size
plt.rcParams['axes.labelsize'] = 12         # Axis label font size

print("✅ All libraries imported successfully!")
print(f"📦 Pandas version: {pd.__version__}")
print(f"📦 NumPy version: {np.__version__}")
print(f"📦 Matplotlib version: {plt.matplotlib.__version__}")
print(f"📦 Seaborn version: {sns.__version__}")

✅ All libraries imported successfully!
📦 Pandas version: 2.0.3
📦 NumPy version: 1.24.3
📦 Matplotlib version: 3.7.2
📦 Seaborn version: 0.12.2


## 2. Data Acquisition

Download the Telco Customer Churn dataset from IBM's GitHub repository and load it into a pandas DataFrame.

In [2]:
# Dataset URL
DATA_URL = 'https://raw.githubusercontent.com/IBM/telco-customer-churn-on-icp4d/master/data/Telco-Customer-Churn.csv'

# Define file path
project_root = Path('/Users/mihiniboteju/churn-prediction-project')
data_raw_path = project_root / 'data' / 'raw'
csv_file_path = data_raw_path / 'Telco-Customer-Churn.csv'

# Create directory if it doesn't exist
data_raw_path.mkdir(parents=True, exist_ok=True)

# Download and load data
try:
    print("📥 Downloading dataset from IBM GitHub repository...")
    df = pd.read_csv(DATA_URL)
    
    # Save to local directory for future use
    df.to_csv(csv_file_path, index=False)
    print(f"✅ Dataset downloaded and saved to: {csv_file_path}")
    print(f"📊 Dataset loaded successfully into DataFrame 'df'")
    
except Exception as e:
    print(f"❌ Error downloading dataset: {e}")
    print("")
    print("📌 Manual download instructions:")
    print(f"   1. Visit: {DATA_URL}")
    print(f"   2. Save the file to: {csv_file_path}")
    print(f"   3. Re-run this cell")
    
    # Try to load from local file if download fails
    if csv_file_path.exists():
        print("")
        print("📂 Found local copy, loading from file...")
        df = pd.read_csv(csv_file_path)
        print("✅ Dataset loaded from local file")

📥 Downloading dataset from IBM GitHub repository...
✅ Dataset downloaded and saved to: /Users/mihiniboteju/churn-prediction-project/data/raw/Telco-Customer-Churn.csv
📊 Dataset loaded successfully into DataFrame 'df'


## 3. Initial Data Inspection

Perform a comprehensive first look at the dataset to understand its structure, size, and content.

### 3.1 Dataset Shape and Size

In [3]:
# Display dataset dimensions
rows, columns = df.shape

print("📏 DATASET DIMENSIONS")
print("=" * 50)
print(f"Total Customers (Rows):    {rows:,}")
print(f"Total Features (Columns):  {columns}")
print(f"Total Data Points:         {rows * columns:,}")
print("=" * 50)

📏 DATASET DIMENSIONS
Total Customers (Rows):    7,043
Total Features (Columns):  21
Total Data Points:         147,903


### 3.2 First Look at the Data

In [6]:
# Display first 10 rows
print("👀 FIRST 10 ROWS OF THE DATASET")
print("=" * 50)
df.head(10)

👀 FIRST 10 ROWS OF THE DATASET


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   
5  9305-CDSKC  Female              0      No         No       8          Yes   
6  1452-KIOVK    Male              0      No        Yes      22          Yes   
7  6713-OKOMC  Female              0      No         No      10           No   
8  7892-POOKP  Female              0     Yes         No      28          Yes   
9  6388-TABGU    Male              0      No        Yes      62          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   
5               Yes     Fiber optic             No           No   
6               Yes     Fiber optic             No          Yes   
7  No phone service             DSL            Yes           No   
8               Yes     Fiber optic             No           No   
9                No             DSL            Yes          Yes   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   
5              Yes          No         Yes             Yes  Month-to-month   
6               No          No         Yes              No  Month-to-month   
7               No          No          No              No  Month-to-month   
8              Yes         Yes         Yes             Yes  Month-to-month   
9               No          No          No              No        One year   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   
5              Yes           Electronic check           99.65        820.5   
6              Yes    Credit card (automatic)           89.10       1949.4   
7               No               Mailed check           29.75        301.9   
8              Yes           Electronic check          104.80      3046.05   
9               No  Bank transfer (automatic)           56.15      3487.95   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  
5   Yes  
6    No  
7    No  
8   Yes  
9    No

### 3.3 Column Names and Data Types

In [7]:
# Display column information
print("📋 COLUMN INFORMATION")
print("=" * 50)
df.info()

📋 COLUMN INFORMATION
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  704

In [8]:
# Detailed data type breakdown
print("\n🔍 DATA TYPE BREAKDOWN")
print("=" * 50)
dtype_counts = df.dtypes.value_counts()
print(dtype_counts)
print("\n")

# List columns by type
print("NUMERICAL COLUMNS:")
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
print(f"  {len(numerical_cols)} columns: {numerical_cols}")
print("\n")

print("CATEGORICAL/OBJECT COLUMNS:")
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()
print(f"  {len(categorical_cols)} columns: {categorical_cols}")


🔍 DATA TYPE BREAKDOWN
object     18
int64       2
float64     1
Name: count, dtype: int64


NUMERICAL COLUMNS:
  3 columns: ['SeniorCitizen', 'tenure', 'MonthlyCharges']


CATEGORICAL/OBJECT COLUMNS:
  18 columns: ['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn']


### 3.4 Missing Values Analysis

In [9]:
# Check for missing values
print("🔎 MISSING VALUES ANALYSIS")
print("=" * 50)

missing_counts = df.isnull().sum()
missing_percentages = (df.isnull().sum() / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing_Count': missing_counts,
    'Missing_Percentage': missing_percentages
})

# Filter to show only columns with missing values
missing_df = missing_df[missing_df['Missing_Count'] > 0].sort_values('Missing_Count', ascending=False)

if len(missing_df) > 0:
    print(missing_df)
else:
    print("✅ No missing values detected in any column!")

print("\n")
print(f"Total missing values: {df.isnull().sum().sum()}")

🔎 MISSING VALUES ANALYSIS
✅ No missing values detected in any column!


Total missing values: 0


### 3.5 Statistical Summary

In [10]:
# Statistical summary of numerical features
print("📊 STATISTICAL SUMMARY - NUMERICAL FEATURES")
print("=" * 50)
df.describe()

📊 STATISTICAL SUMMARY - NUMERICAL FEATURES


SeniorCitizen   tenure  MonthlyCharges
count        7043.00  7043.00         7043.00
mean            0.16    32.37           64.76
std             0.37    24.56           30.09
min             0.00     0.00           18.25
25%             0.00     9.00           35.50
50%             0.00    29.00           70.35
75%             0.00    55.00           89.85
max             1.00    72.00          118.75

In [12]:
# Statistical summary of categorical features
print("📊 STATISTICAL SUMMARY - CATEGORICAL FEATURES")
print("=" * 50)
df.describe(include=['object'])

📊 STATISTICAL SUMMARY - CATEGORICAL FEATURES


customerID gender Partner Dependents PhoneService MultipleLines  \
count         7043   7043    7043       7043         7043          7043   
unique        7043      2       2          2            2             3   
top     7590-VHVEG   Male      No         No          Yes            No   
freq             1   3555    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3473        2810            2785            3875   

       PaperlessBilling     PaymentMethod TotalCharges Churn  
count              7043              7043         7043  7043  
unique                2                 4         6531     2  
top                 Yes  Electronic check                 No  
freq               4171              2365           11  5174

### 3.6 Target Variable Analysis

In [13]:
# Analyze the target variable (Churn)
print("🎯 TARGET VARIABLE ANALYSIS - CHURN")
print("=" * 50)

churn_counts = df['Churn'].value_counts()
churn_percentages = df['Churn'].value_counts(normalize=True) * 100

churn_summary = pd.DataFrame({
    'Count': churn_counts,
    'Percentage': churn_percentages
})

print(churn_summary)
print("\n")

# Calculate churn rate
if 'Yes' in churn_counts:
    churn_rate = (churn_counts['Yes'] / len(df)) * 100
    print(f"📈 Overall Churn Rate: {churn_rate:.2f}%")
    print(f"📊 This means {churn_counts['Yes']:,} out of {len(df):,} customers churned.")
    
    # Class balance assessment
    if churn_rate < 30:
        print("\n⚠️  NOTE: Class imbalance detected (churn rate < 30%).")
        print("   We'll need to address this in modeling phase with:")
        print("   - class_weight='balanced' parameter")
        print("   - Focus on recall and F1-score metrics")

🎯 TARGET VARIABLE ANALYSIS - CHURN
       Count  Percentage
Churn                   
No      5174       73.46
Yes     1869       26.54


📈 Overall Churn Rate: 26.54%
📊 This means 1,869 out of 7,043 customers churned.

⚠️  NOTE: Class imbalance detected (churn rate < 30%).
   We'll need to address this in modeling phase with:
   - class_weight='balanced' parameter
   - Focus on recall and F1-score metrics


### 3.7 Memory Usage

In [ ]:
# Check memory usage
print("💾 MEMORY USAGE")
print("=" * 50)

memory_usage = df.memory_usage(deep=True).sum() / 1024**2  # Convert to MB
print(f"Total memory used by DataFrame: {memory_usage:.2f} MB")
print("\n")
print("Memory usage by column:")
mem_by_col = df.memory_usage(deep=True).sort_values(ascending=False)
mem_by_col_mb = mem_by_col / 1024**2
print(mem_by_col_mb.head(10))

## 4. Data Dictionary

Understanding each feature and its business meaning is crucial for meaningful analysis.

---

### 📋 Feature Categories

#### **A. Customer Demographics (4 features)**

| Feature | Description | Type | Values |
|---------|-------------|------|--------|
| **customerID** | Unique identifier for each customer | Categorical | Unique string |
| **gender** | Customer's gender | Categorical | Male, Female |
| **SeniorCitizen** | Whether customer is 65+ years old | Binary | 0 (No), 1 (Yes) |
| **Partner** | Whether customer has a partner | Categorical | Yes, No |
| **Dependents** | Whether customer has dependents | Categorical | Yes, No |

---

#### **B. Service Information (9 features)**

| Feature | Description | Type | Values |
|---------|-------------|------|--------|
| **PhoneService** | Whether customer has phone service | Categorical | Yes, No |
| **MultipleLines** | Whether customer has multiple phone lines | Categorical | Yes, No, No phone service |
| **InternetService** | Type of internet service | Categorical | DSL, Fiber optic, No |
| **OnlineSecurity** | Whether customer has online security add-on | Categorical | Yes, No, No internet service |
| **OnlineBackup** | Whether customer has online backup add-on | Categorical | Yes, No, No internet service |
| **DeviceProtection** | Whether customer has device protection add-on | Categorical | Yes, No, No internet service |
| **TechSupport** | Whether customer has tech support add-on | Categorical | Yes, No, No internet service |
| **StreamingTV** | Whether customer has streaming TV service | Categorical | Yes, No, No internet service |
| **StreamingMovies** | Whether customer has streaming movies service | Categorical | Yes, No, No internet service |

---

#### **C. Account Information (7 features)**

| Feature | Description | Type | Values/Range |
|---------|-------------|------|-------------|
| **tenure** | Number of months customer has stayed with company | Numerical | 0-72 months |
| **Contract** | Type of contract | Categorical | Month-to-month, One year, Two year |
| **PaperlessBilling** | Whether customer uses paperless billing | Categorical | Yes, No |
| **PaymentMethod** | Customer's payment method | Categorical | Electronic check, Mailed check, Bank transfer, Credit card |
| **MonthlyCharges** | Amount charged to customer monthly | Numerical | $18.25 - $118.75 |
| **TotalCharges** | Total amount charged to customer | Numerical | Continuous |
| **Churn** | **TARGET VARIABLE** - Whether customer left in last month | Categorical | Yes, No |

---

### 🎯 Target Variable: **Churn**

- **Definition:** Whether the customer discontinued service in the last month
- **Values:** 
  - `Yes` = Customer churned (left the company)
  - `No` = Customer retained (still active)
- **Business Importance:** This is what we're trying to predict. Identifying customers likely to churn allows proactive retention efforts.

---

### 💡 Key Business Insights from Features:

1. **Tenure** is likely a strong predictor - longer customers are typically more loyal
2. **Contract Type** may indicate commitment level - month-to-month vs annual contracts
3. **Service bundles** (multiple add-ons) might reduce churn
4. **Payment method** could indicate customer engagement level
5. **Charges** (both monthly and total) represent customer value

---

## 5. Unique Values Inspection

Examine unique values for categorical features to understand the data distribution.

In [ ]:
# Display unique values for each categorical column
print("🔍 UNIQUE VALUES IN CATEGORICAL COLUMNS")
print("=" * 50)

categorical_columns = df.select_dtypes(include=['object']).columns

for col in categorical_columns:
    unique_count = df[col].nunique()
    unique_values = df[col].unique()
    
    print(f"\n{col}:")
    print(f"  Unique count: {unique_count}")
    
    # Only show values if there are reasonable number of unique values
    if unique_count <= 10:
        print(f"  Values: {unique_values}")
    else:
        print(f"  Sample values: {unique_values[:10]}... (showing first 10)")

## 6. Data Quality Issues Identified

Document any issues that need to be addressed in subsequent phases.

In [ ]:
# Check TotalCharges data type issue (common in this dataset)
print("⚠️  DATA QUALITY CHECKS")
print("=" * 50)

# Check if TotalCharges is object type (should be numeric)
if df['TotalCharges'].dtype == 'object':
    print("\n🔴 ISSUE #1: TotalCharges is stored as 'object' instead of numeric")
    print("   This likely means there are non-numeric values.")
    
    # Try to identify the non-numeric values
    try:
        non_numeric = df[pd.to_numeric(df['TotalCharges'], errors='coerce').isnull()]['TotalCharges'].unique()
        print(f"   Non-numeric values found: {non_numeric}")
        print(f"   Count of problematic rows: {df[pd.to_numeric(df['TotalCharges'], errors='coerce').isnull()].shape[0]}")
    except:
        pass
    
    print("   ✅ Solution: Will convert to numeric in Phase 3 (Feature Engineering)")

# Check for duplicate customerIDs
duplicate_ids = df['customerID'].duplicated().sum()
if duplicate_ids > 0:
    print(f"\n🔴 ISSUE #2: Found {duplicate_ids} duplicate customer IDs")
    print("   ✅ Solution: Will investigate and handle in Phase 3")
else:
    print("\n✅ No duplicate customer IDs found")

# Check for potential outliers in numerical columns
print("\n📊 Checking for potential outliers...")
for col in ['tenure', 'MonthlyCharges']:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    outliers = df[(df[col] < (Q1 - 1.5 * IQR)) | (df[col] > (Q3 + 1.5 * IQR))].shape[0]
    print(f"   {col}: {outliers} potential outliers detected")

print("\n" + "=" * 50)

## 7. Initial Observations & Next Steps

### ✅ What We've Accomplished:

1. ✅ Successfully loaded the Telco Customer Churn dataset (7,043 customers × 21 features)
2. ✅ Identified the target variable (Churn) and observed class distribution
3. ✅ Categorized features into demographics, services, and account information
4. ✅ Detected data quality issues (TotalCharges data type)
5. ✅ Established baseline understanding of the dataset structure

---

### 🔍 Key Initial Findings:

1. **Dataset Size:** 7,043 customers with 21 features - sufficient for meaningful ML models
2. **Target Distribution:** Churn rate appears to be ~26% (moderate class imbalance)
3. **Feature Mix:** Good balance of categorical (16) and numerical (3-5) features
4. **Data Quality:** Generally clean, but TotalCharges needs type conversion
5. **No Missing Values:** No explicit NULL values detected (but empty strings may exist)

---

### 🎯 Next Steps (Phase 2 - Full EDA):

1. **Deep Dive Analysis:**
   - Correlation analysis between features and churn
   - Distribution analysis of numerical features
   - Relationship between categorical features and churn

2. **Visualizations to Create:**
   - Churn distribution (bar chart)
   - Tenure vs Churn (histogram/KDE)
   - Monthly Charges vs Total Charges scatter plot
   - Contract Type vs Churn rate (grouped bar chart)
   - Correlation heatmap
   - Services usage patterns

3. **Hypotheses to Test:**
   - Do month-to-month customers churn more than annual contract customers?
   - Is tenure inversely correlated with churn?
   - Do customers with more services have lower churn rates?
   - Does payment method affect churn?

4. **Data Preparation:**
   - Fix TotalCharges data type
   - Identify and handle any remaining data quality issues
   - Prepare data for feature engineering in Phase 3

---

### 📊 Expected Insights:

By the end of Phase 2, we should be able to answer:
- Which features are most strongly correlated with churn?
- What are the characteristics of customers who churn vs those who stay?
- Are there any obvious patterns or segments in the data?
- What features should we focus on for modeling?

---

**Status: Phase 1 Complete ✅**  
**Ready for: Phase 2 - Comprehensive Exploratory Data Analysis**

---